app/src/ui/changes/commit-message.tsx

```diff
@@ -109,112 +104,46 @@ export class CommitMessage extends React.Component<
-    if (nextProps.repository.id !== this.props.repository.id) {
-      this.props.dispatcher.setCommitMessage(this.props.repository, this.state)
-    }
-
+  public componentDidUpdate(prevProps: ICommitMessageProps) {
     if (
-      nextProps.autocompletionProviders !== this.props.autocompletionProviders
+      this.props.autocompletionProviders !== prevProps.autocompletionProviders
     ) {
       this.setState({
         userAutocompletionProvider: findUserAutoCompleteProvider(
-          nextProps.autocompletionProviders
+          this.props.autocompletionProviders
         ),
       })
     }
-
-    // This is rather gnarly. We want to persist the commit message (summary,
-    // and description) in the dispatcher on a per-repository level (git-store).
-    //
-    // Our dispatcher is asynchronous and only emits and update on animation
-    // frames. This is a great thing for performance but it gets real messy
-    // when you throw text boxes into the mix. If we went for a traditional
-    // approach of persisting the textbox values in the dispatcher and updating
-    // the virtual dom when we get new props there's an interim state which
-    // means that the browser can't keep track of the cursor for us, see:
-    //
-    //   http://stackoverflow.com/a/28922465
-    //
-    // So in order to work around that we keep the text values in the component
-    // state. Whenever they get updated we submit the update to the dispatcher
-    // but we disregard the message that flows to us on the subsequent animation
-    // frame unless we have switched repositories.
-    //
-    // Then there's the case when we're being mounted (think switching between
-    // history and changes tabs. In that case we have to rely on what's in the
-    // dispatcher since we don't have any state of our own.
-
-    const nextContextualCommitMessage = nextProps.contextualCommitMessage
-    const lastContextualCommitMessage = this.state.lastContextualCommitMessage
-    // If the contextual commit message changed, we'll use it as our commit
-    // message.
-    if (
-      nextContextualCommitMessage &&
-      (!lastContextualCommitMessage ||
-        !structuralEquals(
-          nextContextualCommitMessage,
-          lastContextualCommitMessage
-        ))
-    ) {
-      this.setState({
-        summary: nextContextualCommitMessage.summary,
-        description: nextContextualCommitMessage.description,
-        lastContextualCommitMessage: nextContextualCommitMessage,
-      })
-    } else if (
-      initializing ||
-      this.props.repository.id !== nextProps.repository.id
-    ) {
-      // We're either initializing (ie being mounted) or someone has switched
-      // repositories. If we receive a message we'll take it
-      if (nextProps.commitMessage) {
-        // Don't update dispatcher here, we're receiving it, could cause never-
-        // ending loop.
+    if (prevProps.commitMessage !== this.props.commitMessage) {
```

Comment:
Should this comparison look at the contents of `commitMessage`, rather than using `===`, to avoid unnecessarily invoking `setState` below?

<img width="321"  src="https://user-images.githubusercontent.com/359239/47799713-9770db00-dd09-11e8-9fd0-071322d41962.png">

Something like `shallowEquals` should do the trick, based on a quick test on my end:

```ts
if (shallowEquals(prevProps.commitMessage, this.props.commitMessage)) {
  return
}
```

Reply:
love it!

Reply:
done in 6b9ca6c